### DISCUSSION TITLE
Get a binding of CGI

### DISCUSSION THREAD

**Discussion by DreepyYunky** [odin-lang/Odin#2433] [Ideas/Requests]
I would like the idea of Common Gateway Interface (CGI) as a possible `vendor` package.

If it already exists, please tell me.

**Comment by Beefster09** [comment]
Wouldn't it make more sense to have an http server package in core? (which is already planned)

CGI has largely fallen out of favor outside of personal websites on university systems. And aside from correctly parsing query strings and form data (which would be part of the http package anyway), it's not that hard to implement yourself by pulling from environment variables and stdin if you _really_ wanted to write a cgi script in Odin.

I feel like you might as well stand up an apache server and run perl scripts if you really want to do CGI. These days, there are just so many easier options for a webserver that I just don't see the appeal to using CGI, especially if you have to deal with all the string conversion manually into a strong, distinct, static type system.

---

You can probably generate some foreign bindings for Odin to whatever c library exists for cgi. I think what we really need is better tools for doing this. There is one that is old and archived linked in the wiki which might serve as a base.

Even still, in a lot of cases, you have to write a fair bit of custom code to make the binding behave nicely and may also need to generate decent wrappers.

I imagine this will be a lot easier with a fully-functioning c parser in core:c rather than just the lexer. Zig may actually be the best tool for the job until then.

**Comment by clsource** [comment]
CGI is still used in some scenarios, but it's not common.

Example

- https://learnbchs.org/

A way to achieve CGI without much hazzle is following these steps:

- Use a language as a Host that has CGI. Example PHP. This will read all the server variables ($_SERVER) and params ($_GET) and pass it as a json string to the odin executable.

- Compile a binary of Odin that can run in shared hosts. Then this binary interprets the odin.file passing the json string as a `-define:env=${JSON}`

You can see an implementation of this workflow for CGI 

In my experimental web framework for Wren: 

- https://github.com/NinjasCL/wren-baron

Here is the same concept applied to executing Clojure in PHP hosts

- https://eccentric-j.com/blog/clojure-like-its-php.html

Keep in mind some shared hosts have old liberary versions, so odin must be compiled to be compatible with such environments.

```
$ ldd --version
ldd (GNU libc) 2.17
```
